File: fsharpapp1/Game.fs
// 推箱子核心逻辑（F# 教学）
module Game

open System.Collections.Generic

type Hist = { Px: int; Py: int; BoxFrom: string option; BoxTo: string option }

type GameState = {
    Walls: HashSet<string>
    Goals: HashSet<string>
    Boxes: HashSet<string>
    mutable Px: int
    mutable Py: int
    mutable Moves: int
    mutable Won: bool
    Width: int
    Height: int
    Hist: ResizeArray<Hist>
}

let key x y = sprintf "%d,%d" x y

let fromRows (rows: string list) =
    let walls = HashSet<string>()
    let goals = HashSet<string>()
    let boxes = HashSet<string>()
    let mutable px, py, maxX, maxY = 0, 0, 0, 0
    rows |> List.iteri (fun y row ->
        maxY <- y
        row |> Seq.iteri (fun x ch ->
            if x > maxX then maxX <- x
            let k = key x y
            match ch with
            | '#' -> walls.Add k |> ignore
            | '.' -> goals.Add k |> ignore
            | '$' -> boxes.Add k |> ignore
            | '*' -> boxes.Add k |> ignore; goals.Add k |> ignore
            | '@' -> px <- x; py <- y
            | '+' -> px <- x; py <- y; goals.Add k |> ignore
            | _ -> ()
        )
    )
    {
        Walls = walls; Goals = goals; Boxes = boxes
        Px = px; Py = py; Moves = 0; Won = false
        Width = maxX + 1; Height = maxY + 1
        Hist = ResizeArray()
    }

let checkWin (s: GameState) =
    s.Won <- s.Boxes |> Seq.forall s.Goals.Contains

let tryMove (s: GameState) dx dy =
    if s.Won then false
    else
        let nx, ny = s.Px + dx, s.Py + dy
        let nk = key nx ny
        if s.Walls.Contains nk then false
        elif s.Boxes.Contains nk then
            let bx, by = nx + dx, ny + dy
            let bk = key bx by
            if s.Walls.Contains bk || s.Boxes.Contains bk then false
            else
                s.Hist.Add { Px = s.Px; Py = s.Py; BoxFrom = Some nk; BoxTo = Some bk }
                s.Boxes.Remove nk |> ignore
                s.Boxes.Add bk |> ignore
                s.Px <- nx; s.Py <- ny
                s.Moves <- s.Moves + 1
                checkWin s
                true
        else
            s.Hist.Add { Px = s.Px; Py = s.Py; BoxFrom = None; BoxTo = None }
            s.Px <- nx; s.Py <- ny
            true

let undo (s: GameState) =
    if s.Won || s.Hist.Count = 0 then false
    else
        let mutable done' = false
        while s.Hist.Count > 0 && not done' do
            let e = s.Hist.[s.Hist.Count - 1]
            s.Hist.RemoveAt(s.Hist.Count - 1)
            match e.BoxFrom with
            | Some bf ->
                s.Px <- e.Px; s.Py <- e.Py
                s.Boxes.Remove e.BoxTo.Value |> ignore
                s.Boxes.Add bf |> ignore
                if s.Moves > 0 then s.Moves <- s.Moves - 1
                s.Won <- false
                done' <- true
            | None ->
                s.Px <- e.Px; s.Py <- e.Py
        true

let renderAscii (s: GameState) =
    let sb = System.Text.StringBuilder()
    for y in 0 .. s.Height - 1 do
        for x in 0 .. s.Width - 1 do
            let k = key x y
            if s.Px = x && s.Py = y then
                sb.Append(if s.Goals.Contains k then '+' else '@') |> ignore
            elif s.Boxes.Contains k then
                sb.Append(if s.Goals.Contains k then '*' else '$') |> ignore
            elif s.Walls.Contains k then sb.Append '#' |> ignore
            elif s.Goals.Contains k then sb.Append '.' |> ignore
            else sb.Append ' ' |> ignore
        sb.AppendLine() |> ignore
    sb.ToString()

File: netaotapp1/Game.cs
// 推箱子核心（.NET Native AOT 教学 — 避免反射/动态特性）
using System.Collections.Generic;
using System.Text;

namespace SokobanAot;

public sealed class GameState
{
    public HashSet<string> Walls { get; } = new();
    public HashSet<string> Goals { get; } = new();
    public HashSet<string> Boxes { get; } = new();
    public int Px { get; private set; }
    public int Py { get; private set; }
    public int Moves { get; private set; }
    public bool Won { get; private set; }
    public int Width { get; private set; }
    public int Height { get; private set; }
    private readonly List<Hist> _hist = new();
    private readonly record struct Hist(int Px, int Py, string? BoxFrom, string? BoxTo, bool IsPush);

    public static string Key(int x, int y) => $"{x},{y}";

    public static GameState FromRows(string[] rows)
    {
        var s = new GameState();
        int maxX = 0, maxY = 0;
        for (int y = 0; y < rows.Length; y++)
        {
            maxY = y;
            for (int x = 0; x < rows[y].Length; x++)
            {
                if (x > maxX) maxX = x;
                var k = Key(x, y);
                switch (rows[y][x])
                {
                    case '#': s.Walls.Add(k); break;
                    case '.': s.Goals.Add(k); break;
                    case '$': s.Boxes.Add(k); break;
                    case '*': s.Boxes.Add(k); s.Goals.Add(k); break;
                    case '@': s.Px = x; s.Py = y; break;
                    case '+': s.Px = x; s.Py = y; s.Goals.Add(k); break;
                }
            }
        }
        s.Width = maxX + 1;
        s.Height = maxY + 1;
        return s;
    }

    public bool TryMove(int dx, int dy)
    {
        if (Won) return false;
        int nx = Px + dx, ny = Py + dy;
        var nk = Key(nx, ny);
        if (Walls.Contains(nk)) return false;
        if (Boxes.Contains(nk))
        {
            int bx = nx + dx, by = ny + dy;
            var bk = Key(bx, by);
            if (Walls.Contains(bk) || Boxes.Contains(bk)) return false;
            _hist.Add(new Hist(Px, Py, nk, bk, true));
            Boxes.Remove(nk);
            Boxes.Add(bk);
            Px = nx; Py = ny;
            Moves++;
            Won = true;
            foreach (var b in Boxes)
                if (!Goals.Contains(b)) { Won = false; break; }
            return true;
        }
        _hist.Add(new Hist(Px, Py, null, null, false));
        Px = nx; Py = ny;
        return true;
    }

    public bool Undo()
    {
        if (Won || _hist.Count == 0) return false;
        while (_hist.Count > 0)
        {
            var e = _hist[_hist.Count - 1];
            _hist.RemoveAt(_hist.Count - 1);
            if (e.IsPush)
            {
                Px = e.Px; Py = e.Py;
                Boxes.Remove(e.BoxTo!);
                Boxes.Add(e.BoxFrom!);
                if (Moves > 0) Moves--;
                Won = false;
                return true;
            }
            Px = e.Px; Py = e.Py;
        }
        return true;
    }

    public string RenderAscii()
    {
        var sb = new StringBuilder();
        for (int y = 0; y < Height; y++)
        {
            for (int x = 0; x < Width; x++)
            {
                var k = Key(x, y);
                if (Px == x && Py == y) sb.Append(Goals.Contains(k) ? '+' : '@');
                else if (Boxes.Contains(k)) sb.Append(Goals.Contains(k) ? '*' : '$');
                else if (Walls.Contains(k)) sb.Append('#');
                else if (Goals.Contains(k)) sb.Append('.');
                else sb.Append(' ');
            }
            sb.AppendLine();
        }
        return sb.ToString();
    }
}
